### ISSUE TITLE
Is session refresh implemented?

### ISSUE THREAD

**Issue by septatrix** [Atinux/nuxt-auth-utils#91]
I saw the `"offline_access"` scope being used for the OAuth0 provider but no reference to refresh tokens in the codebase. Are refresh tokens implemented/utilized? Or is the session from the OAuth2 provider only used once and afterwards everything is delegated to h3?

**Comment by PGLongo** [comment]
@septatrix I have successfully achieved the refresh of the session with the session hook for 'fetch'. If the session has expired and I have a valid refresh token, then the refresh workflow is initiated to obtain a new valid token.

**Comment by septatrix** [comment]
> @septatrix I have successfully achieved the refresh of the session with the session hook for 'fetch'. If the session has expired and I have a valid refresh token, then the refresh workflow is initiated to obtain a new valid token.

Would you mind sharing the code for that?

**Comment by silvio-e** [comment]
@PGLongo I would be also very interested in that! 😊

**Comment by PGLongo** [comment]
Sure! Here I refresh the Microsoft Oauth. Note that in the auth handler I have stored the `expirationDate` in the `session.user`

```js
// server/plugins/session.ts


import { useRuntimeConfig } from '#imports'
import type { OAuthMicrosoftConfig } from '~/server/api/auth/login.get'

export default defineNitroPlugin(() => {
  sessionHooks.hook('fetch', async (session, event) => {
    const now = new Date()
    const expirationDate = new Date(session.user.expirationDate)
    const jwt = getCookie(event, 'jwt')
    console.log(expirationDate < now, expirationDate, now)
    if (expirationDate < now || !jwt) {
      const config = useRuntimeConfig(event).oauth?.microsoft as OAuthMicrosoftConfig

      const tokenEndpoint = `https://login.microsoftonline.com/${config.tenant!}/oauth2/v2.0/token`
      const params = new URLSearchParams()
      const refreshToken = getCookie(event, 'refresh-token') || ''

      params.append('client_id', config.clientId!)
      params.append('client_secret', config.clientSecret!)
      params.append('refresh_token', refreshToken)
      params.append('grant_type', 'refresh_token')

      const data = await $fetch(tokenEndpoint, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded'
        },
        body: params
      })
      const now = new Date()
      session.user.expirationDate = new Date(now.getTime() + data.expires_in * 1000)
      await setCookie(event, 'jwt', data.access_token, { httpOnly: true, secure: true, maxAge: data.expires_in })
    }
  })

  sessionHooks.hook('clear', async (session, event) => {
    await deleteCookie(event, 'jwt')
    await deleteCookie(event, 'refresh-token')
  })
})
```

**Comment by silvio-e** [comment]
@PGLongo Awesome! Very kind of you! Thank you very much!

**Comment by PGLongo** [comment]
> @PGLongo Awesome! Very kind of you! Thank you very much!

Sharing is caring! 😊 

If you have any questions or need further assistance, feel free to reach out. Happy coding!

**Comment by thijsw** [comment]
Thank you, @PGLongo, for providing an example implementation of the refresh dynamic!

Inspired by your code, I created a similar plugin that refreshes the tokens when the access token expires. The problem I'm facing is that the sealed session cookie is never updated, so the original contents remain unchanged. After the access token expires the first time, it refreshes the tokens on every subsequent page refresh. Do you have a solution for this issue?

My code:

```ts
// server/plugins/session.ts

export default defineNitroPlugin(() => {
  sessionHooks.hook('fetch', async (session, event) => {
    const authenticationConfig = getAuthenticationConfig(event) // Configuration helper
    const now = new Date()
    const expirationDate = new Date(session.expirationDate)

    if (expirationDate < now) {
      // Refresh session
      const body = new FormData()
      body.append('grant_type', 'refresh_token')
      body.append('refresh_token', session.refreshToken)
      body.append('response_type', 'id_token')
      body.append('client_id', authenticationConfig.clientId)
      body.append('client_secret', authenticationConfig.clientSecret)
      body.append('scope', authenticationConfig.scope)

      const token = await $fetch<AccessToken>(authenticationConfig.tokenURL, {
        method: 'post',
        body
      })

      session.accessToken = token.access_token
      session.refreshToken = token.refresh_token
      session.expirationDate = new Date(now.getTime() + token.expires_in * 1000)
    }
  })
})
```

**Comment by doubleujay** [comment]
@PGLongo do you have an answer on the question of @thijsw ? We are really looking forward to it.

**Comment by PGLongo** [comment]
@doubleujay and @thijsw are you refreshing the token on server side?

**Comment by thijsw** [comment]
@PGLongo Yes, what I'm trying to accomplish is that the tokens is refreshed server-side (when expired) and the new updated value stored in the encrypted cookie. This new, just retrieved, token should be used by all API calls within the same request cycle. However, using my previous shared code, this doesn't work. The cookie doesn't get updated, so when I refresh the page the old token is being refreshed again. Do you have a suggestion how I could address this?

**Comment by doubleujay** [comment]
@PGLongo Is the reply of @thijsw enough to give me insight?

**Comment by daniandl** [comment]
> Inspired by your code, I created a similar plugin that refreshes the tokens when the access token expires. The problem I'm facing is that the sealed session cookie is never updated, so the original contents remain unchanged. After the access token expires the first time, it refreshes the tokens on every subsequent page refresh. Do you have a solution for this issue?

I can maybe provide some insight here, dealing with this myself at the moment.
The reason the cookie may not be getting refreshed is because the `set-cookie` header is not forwarded on fetch calls.
This eventually lead me to a fetch wrapper snippet like this:
```ts
const res = await $fetch.raw<T>(request, {
  ...opts,
  headers: { ...opts?.headers, ...useRequestHeaders(['cookie']) },
})

// forward cookies into SSR response
const cookies = (res.headers.get('set-cookie') || '').split(',')
for (const cookie of cookies)
  appendResponseHeader(event, 'set-cookie', cookie)

// Return the data of the response
return res._data
```

The cookie now get set refreshed properly for me but pressing back on the browser messes this up and theres an nuxt internal error. 
All this trouble kind of made me rethink my token flow, I do not see a clean refresh token implementation possible on Nuxt, or SSR flows in general, where server side refreshes are needed.
My conclusion here is to just ditch the refresh flow for a regular access token + auto session extension on activity + revocation flow.

Hope it helps a bit at least :)

**Comment by marr** [comment]
Does the `refreshCookie` util help with the stale cookie issue reported by @thijsw and @daniandl? Documentation is here https://nuxt.com/docs/api/utils/refresh-cookie

**Comment by sdevogel** [comment]
Hi, currently also running into the same problem. When I `replaceUserSession` with new experiationDate the cookie is not updated and the session data also stays the same. Any idea how to replace the session when the refresh token is still valid? Thanks!

**Comment by carlos-duran** [comment]
@sdevogel are you calling `await fetch()` from useUserSession() in the client after you replaced the session in the server?

**Comment by sdevogel** [comment]
> @sdevogel are you calling `await fetch()` from useUserSession() in the client after you replaced the session in the server?

No, how would the client know that the server replaced the session? Maybe asking for the obvious but I'm a bit lost. Thanks a lot for your help :)

**Comment by Velka-DEV** [comment]
Can we discuss an implementation design to refresh the token here? We use this library in our application, and it’s not ideal for the session to expire while the user is using the app.

I’m willing to create a pull request after this discussion.

@atinux

**Comment by atinux** [comment]
Well, does this fix your issue?

https://github.com/patrick-hofmann/nuxt-auth-utils/tree/feature/add-server-side-session-storage?tab=readme-ov-file#extending-cookie-lifetime

**Comment by atinux** [comment]
Thank you for your explanation and code examples @haakonmt 

I released https://github.com/atinux/nuxt-auth-utils/releases/tag/v0.5.4 that should improve the `useUserSession().clear()`during SSR.

I created this middleware in the playground to deal with refresh tokens that works with SSR and make sure that the `useFetch` and `useRequestFetch()` requests use the latest cookie: https://github.com/atinux/nuxt-auth-utils/blob/main/playground/middleware/jwt.global.ts

This is how it looks like to apply the updated cookie to the next request happening during the server-side rendering:

```ts
import { parse, parseSetCookie, serialize } from 'cookie-es'

await useRequestFetch()('/api/jtw/refresh', {
  method: 'POST',
  onResponse({ response: { headers } }) {
    // Forward the Set-Cookie header to the main server event
    if (import.meta.server && serverEvent) {
      for (const setCookie of headers.getSetCookie()) {
        appendResponseHeader(serverEvent, 'Set-Cookie', setCookie)
        // Update session cookie for next fetch requests
        const { name, value } = parseSetCookie(setCookie)
        if (name === runtimeConfig.session.name) {
          // console.log('updating headers.cookie to', value)
          const cookies = parse(serverEvent.headers.get('cookie') || '')
          // set or overwrite existing cookie
          cookies[name] = value
          // update cookie event header for future requests
          serverEvent.headers.set('cookie', Object.entries(cookies).map(([name, value]) => serialize(name, value)).join('; '))
          // Also apply to serverEvent.node.req.headers
          if (serverEvent.node?.req?.headers) {
            serverEvent.node.req.headers['cookie'] = serverEvent.headers.get('cookie') || ''
          }
        }
      }
    }
  },
})
```

The only solution I can see for you to avoid doing this is for `nuxt-auth-utils` to create:
- `useAuthFetch()` composable to replace `useFetch`
- `$authFetch` util to replace `$fetch`

It will have another benefit, being able to know when an API request update the session cookie to automatically refresh the user session on the app-side.

**Comment by pablodz** [comment]
so, what's the best approach?

**Comment by colhugs230** [comment]
Did anything ever come of this?

**Comment by ArbazIrshad** [comment]
@atinux Is there a plan to add this feature to the package? If no, then what is best approach to handle this?
Maybe mentioning this in the docs will also be a good idea.
